src/gui/history/qgshistoryentrynode.h

```diff
@@ -0,0 +1,191 @@
+/***************************************************************************
+                            qgshistoryentrynode.h
+                            --------------------------
+    begin                : Aprial 2023
+    copyright            : (C) 2023 by Nyall Dawson
+    email                : nyall dot dawson at gmail dot com
+ ***************************************************************************/
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#ifndef QGSHISTORYENTRYNODE_H
+#define QGSHISTORYENTRYNODE_H
+
+#include "qgis_gui.h"
+#include "qgis_sip.h"
+#include "qgis.h"
+#include <QString>
+#include <QVariantMap>
+#include <memory>
+#include <deque>
+
+class QWidget;
+class QAction;
+class QgsHistoryEntryGroup;
+class QgsHistoryWidgetContext;
+
+/**
+ * Base class for nodes representing a QgsHistoryEntry.
+ *
+ * \ingroup gui
+ * \since QGIS 3.32
+ */
+class GUI_EXPORT QgsHistoryEntryNode
+{
+  public:
+
+    /**
+     * Constructor for QgsHistoryEntryNode.
+     */
+    QgsHistoryEntryNode() = default;
+    virtual ~QgsHistoryEntryNode();
+
+    //! QgsHistoryEntryNode cannot be copied
+    QgsHistoryEntryNode( const QgsHistoryEntryNode &other ) = delete;
+    //! QgsHistoryEntryNode cannot be copied
+    QgsHistoryEntryNode &operator=( const QgsHistoryEntryNode &other ) = delete;
+
+    /**
+     * Returns the node's parent node.
+     *
+     * If parent is NULLPTR, the node is a root node.
+     */
+    QgsHistoryEntryGroup *parent() { return mParent; }
+
+    /**
+     * Returns the node's data for the specified model \a role.
+     */
+    virtual QVariant data( int role = Qt::DisplayRole ) const = 0;
+
+    /**
+     * Returns the number of child nodes owned by this node.
+     */
+    virtual int childCount() const;
+
+    /**
+     * Returns a HTML formatted text string which should be shown to a user when
+     * selecting the node.
+     *
+     * Subclasses should implement this method or createWidget(), but not both.
+     */
```

Comment:
Maybe adding a \see createWidget() here (and a \see html() in the createWidget doc), it'll make it easier to jump back and forth in the documentation website.